cmd/thanos/store.go

```diff
@@ -106,104 +110,106 @@ func runStore(
-
-			err := runutil.Repeat(syncInterval, ctx.Done(), func() error {
-				if err := bs.SyncBlocks(ctx); err != nil {
-					level.Warn(logger).Log("msg", "syncing blocks failed", "err", err)
-				}
-				return nil
-			})
-
-			runutil.CloseWithLogOnErr(logger, bs, "bucket store")
-			return err
-		}, func(error) {
-			cancel()
-		})
+	indexCache, err := storecache.NewIndexCache(logger, reg, storecache.Opts{
+		MaxSizeBytes:     indexCacheSizeBytes,
+		MaxItemSizeBytes: maxItemSizeBytes,
+	})
+	if err != nil {
+		return errors.Wrap(err, "create index cache")
+	}
 
-		l, err := net.Listen("tcp", grpcBindAddr)
-		if err != nil {
-			return errors.Wrap(err, "listen API address")
-		}
+	bs, err := store.NewBucketStore(
+		logger,
+		reg,
+		bkt,
+		dataDir,
+		indexCache,
+		chunkPoolSizeBytes,
+		maxSampleCount,
+		maxConcurrent,
+		verbose,
+		blockSyncConcurrency,
+		filterConf,
+	)
+	if err != nil {
+		return errors.Wrap(err, "create object storage store")
+	}
 
-		opts, err := defaultGRPCServerOpts(logger, reg, tracer, cert, key, clientCA)
-		if err != nil {
-			return errors.Wrap(err, "grpc server options")
-		}
+	begin := time.Now()
+	level.Debug(logger).Log("msg", "initializing bucket store")
+	if err := bs.InitialSync(context.Background()); err != nil {
+		return errors.Wrap(err, "bucket store initial sync")
+	}
+	level.Debug(logger).Log("msg", "bucket store ready", "init_duration", time.Since(begin).String())
 
-		s := grpc.NewServer(opts...)
-		storepb.RegisterStoreServer(s, bs)
+	ctx, cancel := context.WithCancel(context.Background())
+	g.Add(func() error {
+		defer runutil.CloseWithLogOnErr(logger, bkt, "bucket client")
 
-		g.Add(func() error {
-			level.Info(logger).Log("msg", "Listening for StoreAPI gRPC", "address", grpcBindAddr)
-			return errors.Wrap(s.Serve(l), "serve gRPC")
-		}, func(error) {
-			s.Stop()
+		err := runutil.Repeat(syncInterval, ctx.Done(), func() error {
+			if err := bs.SyncBlocks(ctx); err != nil {
+				level.Warn(logger).Log("msg", "syncing blocks failed", "err", err)
+			}
+			return nil
 		})
-	}
-	if err := metricHTTPListenGroup(g, logger, reg, httpBindAddr); err != nil {
+
+		runutil.CloseWithLogOnErr(logger, bs, "bucket store")
 		return err
+	}, func(error) {
+		cancel()
+	})
+
+	l, err := net.Listen("tcp", grpcBindAddr)
+	if err != nil {
+		return errors.Wrap(err, "listen API address")
+	}
+
+	opts, err := defaultGRPCServerOpts(logger, reg, tracer, cert, key, clientCA)
+	if err != nil {
+		return errors.Wrap(err, "grpc server options")
+	}
+
+	s := grpc.NewServer(opts...)
+	storepb.RegisterStoreServer(s, bs)
+
+	g.Add(func() error {
+		level.Info(logger).Log("msg", "Listening for StoreAPI gRPC", "address", grpcBindAddr)
+		statusProber.SetReady()
+		return errors.Wrap(s.Serve(l), "serve gRPC")
+	}, func(error) {
+		s.Stop()
+	})
+
+	if err := defaultHTTPListener(g, logger, reg, httpBindAddr, statusProber); err != nil {
+		return errors.Wrap(err, "create readiness prober")
```

Comment:
technically we create HTTP Server not prober

Reply:
you are right, changed to `scheduling HTTP server`